### ISSUE TITLE
Generic defaults break TS inference when wrapping queries

### ISSUE THREAD

**Issue by g1eny0ung** [anymaniax/orval#694]
## What are the steps to reproduce this issue?

1. With React Query generated, I want to wrap some queries with some default options, like:

    ```typescript
    export const useGetCommonConfig = <
      TData = Awaited<ReturnType<typeof getCommonConfig>>,
      TError = UtilsAPIError
    >(options?: {
      query?: UseQueryOptions<Awaited<ReturnType<typeof getCommonConfig>>, TError, TData>
    }): UseQueryResult<TData, TError> & { queryKey: QueryKey } => {
    // ...
    }

    export const useCustomGetCommonConfig = (...args: Parameters<typeof useGetCommonConfig>) => {
      return useGetCommonConfig({
        query: {
          staleTime: Stale.DAY,
          ...args[0]?.query,
        },
      })
   }
    ```
2. But this will cause the `TData` type return by the wrapper to be `unknown`:

   <img width="483" alt="image" src="https://user-images.githubusercontent.com/15034155/206420700-83cfb2f7-66f7-4159-97d3-baaa3fd3f75d.png">

3. Since the `TData` generic type is no constraint but accepts a default type, the TS compiler cannot infer the final type from the parameters if the generic type is not filled in manually.

## What happens?

See above.

## What were you expecting to happen?

**Use `TData extends Awaited<ReturnType<typeof getCommonConfig>>` rather than `TData = Awaited<ReturnType<typeof getCommonConfig>>`**, this also prevents the user from filling in some wrong type, if there is no constraint then the user can use an unintended type such as `number` type here.

## Any logs, error output, etc?

…

## Any other comments?

@anymaniax I'm willing to help improve this if that's okay with you. 😃

You can also find the actual code related to this issue at https://github.com/chaos-mesh/chaos-mesh/pull/3748/files#diff-3a5ef0dc8a96c626a60dfa618f5c93bbfcc8e053fd66e98dc3e88ac21dbe1030.

## What versions are you using?

**Operating System:** …
**Package Version:** `v6.10.3`
**Browser Version:** …

**Comment by anymaniax** [comment]
Hello @g1eny0ung, with `orval@alpha` you can now provide a queryOptions mutator in the orval config. Like that you can quickly provide some shared options. I don't know if it would solve your problem.

**Comment by anymaniax** [comment]
and It's not an extend because otherwise, the options `select` would break if you return something that doesn't extend that type.

**Comment by g1eny0ung** [comment]
> Hello @g1eny0ung, with `orval@alpha` you can now provide a queryOptions mutator in the orval config. Like that you can quickly provide some shared options. I don't know if it would solve your problem.

Thanks. I will take it a try.

**Comment by g1eny0ung** [comment]
> and It's not an extend because otherwise, the options `select` would break if you return something that doesn't extend that type.

Indeed. 😮 I missed that React Query also has a `select` option, thanks for your help. I will now close this issue.
